sympy/simplify/cse_main.py

```diff
@@ -136,6 +137,235 @@ def postprocess_for_cse(expr, optimizations):
+        """
+        for arg in self.func_to_argset[func_i]:
+            self.arg_to_funcset[arg].remove(func_i)
+
+    def get_common_arg_candidates(self, argset, min_func_i, threshold=2):
+        """
+        Return a dict whose keys are function numbers. The entries of the dict are
+        the number of arguments said function has in common with `argset`. Entries
+        have at least `threshold` items in common.
+        """
+        from collections import defaultdict
+        count_map = defaultdict(lambda: 0)
+
+        # Sorted by size to make best use of the performance hack below.
+        funcsets = sorted((self.arg_to_funcset[arg] for arg in argset), key=len)
+
+        for funcset in funcsets[:-threshold+1]:
+            for func_i in funcset:
+                if func_i >= min_func_i:
+                    count_map[func_i] += 1
+
+        for i, funcset in enumerate(funcsets[-threshold+1:]):
+            # When looking at the tail end of the funcsets list, items below
+            # this threshold in the count_map don't have to be considered
+            # because they can't possibly be in the output.
+            count_map_threshold = i + 1
+
+            # We pick the smaller of the two containers to iterate over to
+            # reduce the number of items we have to look at.
+            (smaller_funcs_container,
+             larger_funcs_container) = sorted([funcset, count_map], key=len)
+
+            for func_i in smaller_funcs_container:
+                if count_map[func_i] < count_map_threshold:
+                    continue
+
+                if func_i in larger_funcs_container:
+                    count_map[func_i] += 1
+
+        return dict(
+            (k, v) for k, v in count_map.items()
+            if v >= threshold)
+
+    def get_subset_candidates(self, argset, restrict_to_funcset=None):
+        """
+        Return a set of functions each of which whose argument list contains
+        `argset`, optionally filtered only to contain functions in
+        `restrict_to_funcset`.
+        """
+        iarg = iter(argset)
+
+        indices = set(
+            fi for fi in self.arg_to_funcset[next(iarg)])
+
+        if restrict_to_funcset is not None:
+            indices &= restrict_to_funcset
+
+        for arg in iarg:
+            indices &= self.arg_to_funcset[arg]
+
+        return indices
+
+    def update_func_argset(self, func_i, new_argset):
+        """
+        Update a function with a new set of arguments.
+        """
+        new_args = set(new_argset)
+        old_args = self.func_to_argset[func_i]
+
+        for deleted_arg in old_args - new_args:
+            self.arg_to_funcset[deleted_arg].remove(func_i)
+        for added_arg in new_args - old_args:
+            self.arg_to_funcset[added_arg].add(func_i)
+
+        self.func_to_argset[func_i].clear()
+        self.func_to_argset[func_i].update(new_args)
+
+
+class Unevaluated(object):
```

Comment:
This was initially added for symengine compatibility in sumpy, but this also has a performance benefit in that we can identify which expressions are needed to be evaluated at the end of the cse. (https://github.com/sympy/sympy/pull/13220/files#diff-44a512e7d608776ee035c0ab6a4baa75L411 was removed because of this)